File: core/scheduler.go
package core

import (
	"zephyria/state"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
)

// Wave contains a set of transactions that can be executed in parallel.
type Wave struct {
	Transactions []*types.Transaction
}

// schedule groups transactions into waves using a refined dependency analysis.
// It considers:
// 1. Account-level conflicts (Nonce, Balance)
// 2. Storage-level conflicts (AccessList slots)
// 3. Aquarius Program Lookups
func (e *Executor) schedule(txs []*types.Transaction, statedb *state.StateDB) [][]*types.Transaction {
	if len(txs) == 0 {
		return nil
	}

	signer := types.LatestSigner(e.config)
	var waves [][]*types.Transaction

	// Map to track current writes/reads in the active wave
	writes := make(map[string]bool)
	reads := make(map[string]bool)

	var currentWave []*types.Transaction
	var deferred []*types.Transaction

	queue := txs
	for len(queue) > 0 {
		for _, tx := range queue {
			sender, _ := types.Sender(signer, tx)
			recipient := common.Address{}
			if tx.To() != nil {
				recipient = *tx.To()
			}

			// Identify Write/Read Sets
			txWrites := make(map[string]bool)
			txReads := make(map[string]bool)

			// 1. Account Level Dependencies
			// Sender: Nonce and Balance are ALWAYS written (gas, etc)
			txWrites[string(state.NonceKey(sender))] = true
			txWrites[string(state.BalanceKey(sender))] = true

			// 4. Automatic Data Account Resolution (Aquarius Auto-Binding)
			// Always Shard Programs to Data Accounts (Private Mode).
			// This matches Executor behavior. Access Lists for the Recipient are remapped to Data Shard.

			if tx.To() != nil {
				target := recipient // Note: recipient is defined as *tx.To() above

				// Optimization: Only check if it has code (IsProgramAccount)
				if statedb.IsProgramAccount(target) {
					// It's a contract. In Aquarius, we write to the User's Data Shard.
					dataAddr := state.DeriveDataAddress(sender, target)

					// Dependency: The Data Account is WRITTEN (Balance, Storage, etc)
					txWrites[string(state.AccountStem(dataAddr))] = true

					// Dependency: The Program is READ ONLY
					txReads[string(state.ProgramKey(target))] = true
					txReads[string(state.CodeHashKey(target))] = true
				} else {
					// EOA or Precompile (or empty)
					if tx.Value().Sign() > 0 {
						txWrites[string(state.BalanceKey(target))] = true
					}
				}
			}

			// 2. Storage Level Dependencies (AccessList)
			// Crucial: If AL targets the *Recipient*, we must lock the *Data Account* (if sharded),
			// NOT the Contract Address (which would force serialization).
			for _, entry := range tx.AccessList() {
				lockAddr := entry.Address

				// Remap logic: If entry is for Recipient contract, it refers to Data Shard
				if tx.To() != nil && entry.Address == recipient && statedb.IsProgramAccount(recipient) {
					lockAddr = state.DeriveDataAddress(sender, recipient)
				}

				if len(entry.StorageKeys) == 0 {
					txWrites[string(state.AccountStem(lockAddr))] = true
					txWrites[string(state.BalanceKey(lockAddr))] = true
					txWrites[string(state.NonceKey(lockAddr))] = true
				} else {
					for _, slot := range entry.StorageKeys {
						txWrites[string(state.StorageKey(lockAddr, slot))] = true
					}
				}
			}

			// Check Conflicts with current wave
			conflict := false
			for w := range txWrites {
				if writes[w] || reads[w] {
					conflict = true
					break
				}
			}
			if !conflict {
				for r := range txReads {
					if writes[r] {
						conflict = true
						break
					}
				}
			}

			if !conflict {
				// No conflict, add to current wave
				currentWave = append(currentWave, tx)
				for w := range txWrites {
					writes[w] = true
				}
				for r := range txReads {
					reads[r] = true
				}
			} else {
				// Deferred to next wave
				deferred = append(deferred, tx)
			}
		}

		if len(currentWave) == 0 && len(deferred) > 0 {
			// Emergency: take first deferred to prevent deadlock
			currentWave = append(currentWave, deferred[0])
			deferred = deferred[1:]
		}

		waves = append(waves, currentWave)

		// Reset for next wave
		queue = deferred
		deferred = nil
		currentWave = nil
		writes = make(map[string]bool)
		reads = make(map[string]bool)
	}

	return waves
}

File: src/core/scheduler.zig
const std = @import("std");
const types = @import("types.zig");
const state = @import("state.zig");

pub const Scheduler = struct {
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) Scheduler {
        return Scheduler{
            .allocator = allocator,
        };
    }

    /// Groups transactions into waves that can be executed in parallel.
    /// Each wave contains transactions that do not have conflicting Read/Write sets.
    pub fn schedule(self: *Scheduler, txs: []*types.Transaction, state_obj: *state.State) ![][]*types.Transaction {
        if (txs.len == 0) return try self.allocator.alloc([]*types.Transaction, 0);

        var waves = std.ArrayListUnmanaged([]*types.Transaction){};
        errdefer {
            for (waves.items) |w| self.allocator.free(w);
            waves.deinit(self.allocator);
        }

        // Global tracked reads/writes for the ACTIVE wave
        var global_writes = std.StringHashMap(void).init(self.allocator);
        defer self.free_kv(&global_writes);
        var global_reads = std.StringHashMap(void).init(self.allocator);
        defer self.free_kv(&global_reads);

        var current_wave = std.ArrayListUnmanaged(*types.Transaction){};
        defer current_wave.deinit(self.allocator);

        // Use a list to track transactions that need scheduling
        var pending = std.ArrayListUnmanaged(*types.Transaction){};
        defer pending.deinit(self.allocator);
        try pending.appendSlice(self.allocator, txs);

        var next_pending = std.ArrayListUnmanaged(*types.Transaction){};
        defer next_pending.deinit(self.allocator);

        while (pending.items.len > 0) {
            for (pending.items) |tx| {
                const sender = tx.from;
                const recipient = tx.to orelse types.Address.zero();

                // 1. Identify local Read/Write sets for this transaction
                var tx_writes = std.ArrayListUnmanaged([32]u8){};
                defer tx_writes.deinit(self.allocator);
                var tx_reads = std.ArrayListUnmanaged([32]u8){};
                defer tx_reads.deinit(self.allocator);

                // Senders always write Nonce and Balance (for gas payment)
                try tx_writes.append(self.allocator, state.State.nonce_key(sender));
                try tx_writes.append(self.allocator, state.State.balance_key(sender));

                if (tx.to) |target| {
                    if (state_obj.is_program_account(target)) {
                        // Aquarius: It's a contract. We write to the User's Data Shard.
                        // For parity, we derive the data address and lock its account stem.
                        const data_addr = self.derive_data_address(sender, target);
                        try tx_writes.append(self.allocator, state.State.balance_key(data_addr));

                        // We READ from the program code
                        try tx_reads.append(self.allocator, state.State.code_hash_key(target));
                    } else if (tx.value > 0) {
                        try tx_writes.append(self.allocator, state.State.balance_key(recipient));
                    }
                }

                // 2. Conflict Detection
                var conflict = false;
                for (tx_writes.items) |w| {
                    if (global_writes.contains(&w) or global_reads.contains(&w)) {
                        conflict = true;
                        break;
                    }
                }
                if (!conflict) {
                    for (tx_reads.items) |r| {
                        if (global_writes.contains(&r)) {
                            conflict = true;
                            break;
                        }
                    }
                }

                if (!conflict) {
                    // Accept into wave
                    try current_wave.append(self.allocator, tx);
                    for (tx_writes.items) |w| try global_writes.put(try self.allocator.dupe(u8, &w), {});
                    for (tx_reads.items) |r| try global_reads.put(try self.allocator.dupe(u8, &r), {});
                } else {
                    try next_pending.append(self.allocator, tx);
                }
            }

            if (current_wave.items.len == 0 and next_pending.items.len > 0) {
                // Deadlock safety: if NO transactions could be scheduled,
                // take the first one serial.
                const forced = next_pending.orderedRemove(0);
                try current_wave.append(self.allocator, forced);
            }

            // Finalize current wave
            try waves.append(self.allocator, try current_wave.toOwnedSlice(self.allocator));

            // Prepare for next wave
            self.free_kv(&global_writes);
            self.free_kv(&global_reads);

            pending.clearRetainingCapacity();
            try pending.appendSlice(self.allocator, next_pending.items);
            next_pending.clearRetainingCapacity();
        }

        return try waves.toOwnedSlice(self.allocator);
    }

    /// Aquarius helper: derives a deterministic data address for (User, Program)
    pub fn derive_data_address(self: *Scheduler, user: types.Address, program: types.Address) types.Address {
        _ = self;
        var h: [32]u8 = undefined;
        var hasher = std.crypto.hash.sha3.Keccak256.init(.{});
        hasher.update(&user.bytes);
        hasher.update(&program.bytes);
        hasher.final(&h);

        var addr: types.Address = undefined;
        @memcpy(&addr.bytes, h[12..32]); // Take last 20 bytes
        return addr;
    }

    fn free_kv(self: *Scheduler, map: *std.StringHashMap(void)) void {
        var it = map.iterator();
        while (it.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
        }
        map.clearRetainingCapacity();
    }
};
